### DISCUSSION TITLE
How to `.defineRule` in `ESLint` class?

### DISCUSSION THREAD

**Discussion by fisker** [eslint/eslint#16333] [General]
When use `ESLint` class like this

```js
new ESLint({plugins: {foo}}).lintFiles('.')
```

The result message missing `ruleId`.

There was [`.defineRule` in `Linter` class](https://eslint.org/docs/latest/developer-guide/nodejs-api#linterdefinerule), how do I give a rule name in `ESLint` class? Or shouldn't it use id from `foo.rules`?

**Comment by fisker** [comment]
I guess I made a mistake, it worked https://github.com/eslint/eslint/blob/c6ee180fd19aad5b8415df2120f65fd7d299d29c/tests/lib/eslint/eslint.js#L2125. Will dig deeper.

**Comment by mdjermanovic** [comment]
> There was [`.defineRule` in `Linter` class](https://eslint.org/docs/latest/developer-guide/nodejs-api#linterdefinerule), how do I give a rule name in `ESLint` class? Or shouldn't it use id from `foo.rules`?

It's using id from `foo.rules`. You're probably missing `plugins: ["foo"]` in the config. Top level `plugins` is to pass preloaded plugin implementations, `config.plugins` is to specify which plugins should be used.

For example:

```js
new ESLint({
    plugins: {
        foo
    },
    baseConfig: {
        rules: {
            "foo/my-rule": 2
        },
        plugins: ["foo"] // <--
    }
});
```

**Comment by fisker** [comment]
No, my real problem is my rule fix to invalid code, then eslint parse the output again, but in the fatal message there is no ruleId, I thought it's missing. As I understand ESLint will merge fixes before the next round of running, so I don't think ESLint can trace which rule cause the broken fix? Context https://github.com/sindresorhus/eslint-plugin-unicorn/pull/1906#issuecomment-1252494853

**Reply by mdjermanovic** [reply]
That's correct. ESLint applies all fixes from all rules and then runs another round of parsing, linting, and fixing. If there is a parsing error in the fixed code, ESLint doesn't trace which rule has caused it, as it doesn't parse code after each individual fix.

**Reply by fisker** [reply]
Thank you for the information.
